README.md
Laravel Response Formatter
=====================================

Small library to send a proper safe JSON `)]}',\n` format response with Laravel.

## Sample

Returning safe json preceded with `)]}',\n` formatted response:

```php
<?php

use AdrianoRosa\HttpResponse\Response;

$data = ['foo' => 'bar'];

$response = Response::create()->safeJson($data);

// )]}',\n
// {"code":200,"status":"success","data":{"foo":"bar"}}
```

Returning default json formatted response:

```php
<?php

use AdrianoRosa\HttpResponse\Response;

$data = ['foo' => 'bar'];

$response = Response::create()->toJson($data);

// {"code":200,"status":"success","data":{"foo":"bar"}}
```

## Version 0.2.x
Compatible with Laravel 5.3

Installation

    composer require adrianorosa/http-response-formatter
    
## Version 0.1.x
Compatible with Laravel 5.2

Installation

    composer require adrianorosa/http-response-formatter:0.1.*

### License
This software is licensed under the MIT License. Please read LICENSE 
for information on the software availability and distribution.


CHANGELOG.md
v0.3.0
------
- requirements upgrade to Laravel v5.4 

v0.2.1
------
- minor refactoring
- refactor `Response::toJson()` and `Response::safeJson()`
- add `Response::jsonError()` and `Response::safeJsonError()`
- add testing append data
- add callStatic `Response::json()`, `Response::sjson()`, `Response::error()`, `Response::serror()` methods 

v0.2.0
------
- add phpunit test case
- add `Response::json()` method to allow send default JSON response format 
- requirements now the library requires PHP >= v5.6 
- requirements upgrade to Laravel v5.3
- removed deprecated method `ResponseFormatter::setAppendToResponse()` in favor of `ResponseFormatter::add()`
- removed deprecated method `ResponseFormatter::getResponseData()`
- removed deprecated method `ResponseFormatter::getResponseData()`
- removed deprecated method `ResponseFormatter::getStatusCode()`
- removed deprecated method `ResponseFormatter::getStatus()`

v0.1.0
-----
- First release


tests/TestCase.php
<?php

use PHPUnit_Framework_Assert as PHPUnit;

class TestCase extends PHPUnit_Framework_TestCase
{
    /**
     * Validate AngularJS Security JSON response
     *
     * @param mixed $response
     *
     * @return void
     */
    public function assertAngularJsonResponse($response = null)
    {
        if ($response instanceof \Illuminate\Http\Response) {
            $response = $response->getContent();
        }

        return PHPUnit::assertRegExp("/^\)\]\}\'\,\n/", $response, 'Invalid Angular json secure response');
    }

    /**
     * Converts the safeJson like response into an array.
     *
     * @param \Illuminate\Http\Response $response
     * @param bool|false $as_string
     *
     * @return \Illuminate\Http\Response|mixed
     */
    protected function fromSafeJson($response = null, $as_string = false)
    {
        if ($response instanceof \Illuminate\Http\JsonResponse) {
            return $response;
        }

        $json = str_replace(")]}',\n", '', $response->getContent());
        return $as_string ? $json : json_decode($json, true);
    }
}


tests/ResponseTest.php
<?php

use AdrianoRosa\HttpResponse\Response;

/**
 * @project: HttpResponseFormatter
 * @file   : ResponseTest.php
 * @author : Adriano Rosa <http://adrianorosa.com>
 * @date   : 2016-12-27 14:52
 */
class ResponseTest extends TestCase
{

    public function testResponseToJson()
    {
        $data = [
          'foo' => 'bar'
        ];

        $response = Response::create()->toJson($data);
        $content = json_decode($response->getContent(), true);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals('{"code":200,"status":"success","data":{"foo":"bar"}}', $response->getContent());

        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);

        // Test for append data
        $response = Response::create()->with('Music', ['Jazz', 'Pop', 'Rock'])->toJson($data);
        $content = json_decode($response->getContent(), true);

        $this->assertEquals('{"code":200,"status":"success","data":{"foo":"bar"},"Music":["Jazz","Pop","Rock"]}', $response->getContent());
        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);
        $this->assertArrayHasKey('Music', $content);
    }

    public function testResponseSafeJson()
    {
        $data = [
            'foo' => 'bar'
        ];

        $response = Response::create()->safeJson($data);
        $content = $this->fromSafeJson($response);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);

        $this->assertEquals(
            '{"code":200,"status":"success","data":{"foo":"bar"}}',
            $this->fromSafeJson($response, true)
        );
        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);

        // Test for append data
        $response = Response::create()->with('Music', ['Jazz', 'Pop', 'Rock'])->safeJson($data);
        $content = $this->fromSafeJson($response);

        $this->assertEquals(
            '{"code":200,"status":"success","data":{"foo":"bar"},"Music":["Jazz","Pop","Rock"]}',
            $this->fromSafeJson($response, true)
        );
        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);
        $this->assertArrayHasKey('Music', $content);
    }

    public function testResponseToJsonPayloadString()
    {
        $data = 'foo';

        $response = Response::create()->toJson($data);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals('{"code":200,"status":"success","data":["foo"]}', $response->getContent());

        $content = json_decode($response->getContent(), true);

        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
    }

    public function testResponseSafeJsonPayloadString()
    {
        $data = 'foo';

        $response = Response::create()->safeJson($data);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals('{"code":200,"status":"success","data":["foo"]}', $this->fromSafeJson($response, true));

        $content = $this->fromSafeJson($response);

        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
    }

    public function testResponseToJsonSTATUS()
    {
        $data = [
            'foo' => 'bar'
        ];

        $response = Response::create()->toJson($data, 301);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(301, $response->status());

        $response = Response::create()->toJson($data, 404);
        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(404, $response->status());

        $response = Response::create()->toJson($data, 200, ['X-Foo' => 'Bar']);
        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(200, $response->status());
        $this->assertArrayHasKey('x-foo', $response->headers->all());
        $this->assertEquals('Bar', $response->headers->get('x-foo'));
    }

    public function testResponseSafeJsonSTATUS()
    {
        $data = [
            'foo' => 'bar'
        ];

        $response = Response::create()->safeJson($data, 301);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(301, $response->status());

        $response = Response::create()->safeJson($data, 404);
        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(404, $response->status());

        $response = Response::create()->safeJson($data, 200, ['X-Foo' => 'Bar']);
        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals(200, $response->status());
        $this->assertArrayHasKey('x-foo', $response->headers->all());
        $this->assertEquals('Bar', $response->headers->get('x-foo'));
    }

    public function testResponseJSONError()
    {
        $response = Response::create()->jsonError('No status code defined');
        $data = json_decode($response->getContent(), true);
        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals('No status code defined', $data['data']['errorDescription']);

        $response = Response::create()->jsonError('Not Found', 404);
        $this->assertEquals(404, $response->getStatusCode());

        $response = Response::create()->jsonError('Generic error code defined', 4001)->setStatusCode(401);
        $this->assertEquals(401, $response->getStatusCode());
        $this->assertEquals(
            '{"code":200,"status":"success","data":{"errorCode":4001,"errorDescription":"Generic error code defined"}}',
            $response->getContent()
        );
    }

    public function testResponseSafeJSONError()
    {
        $response = Response::create()->safeJsonError('No status code defined');
        $data = $this->fromSafeJson($response);

        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals('No status code defined', $data['data']['errorDescription']);

        $response = Response::create()->safeJsonError('Not Found', 404);
        $this->assertEquals(404, $response->getStatusCode());

        $response = Response::create()->safeJsonError('Generic error code defined', 4001)->setStatusCode(401);
        $this->assertEquals(401, $response->getStatusCode());
        $this->assertAngularJsonResponse($response->getContent());
    }

    public function testStaticCalls()
    {
        $data = ['foo' => 'bar'];
        $response = Response::json($data);
        $content = json_decode($response->getContent(), true);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertEquals('{"code":200,"status":"success","data":{"foo":"bar"}}', $response->getContent());

        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);

        // From safeJson
        $data = ['foo' => 'bar'];
        $response = Response::sjson($data);
        $content = $this->fromSafeJson($response);

        $this->assertInstanceOf('AdrianoRosa\HttpResponse\Response', $response);
        $this->assertAngularJsonResponse($response);

        $this->assertArrayHasKey('code', $content);
        $this->assertArrayHasKey('status', $content);
        $this->assertArrayHasKey('data', $content);
        $this->assertArrayHasKey('foo', $content['data']);

        // Error
        $response = Response::error('error description');
        $content = json_decode($response->getContent(), true);
        $this->assertEquals('error description', $content['data']['errorDescription']);
        $this->assertEquals(200, $response->getStatusCode());

        $response = Response::error('error description')->setStatusCode(405);
        $this->assertEquals(405, $response->getStatusCode());

        // safeError
        $response = Response::serror('error description');
        $content = $this->fromSafeJson($response);
        $this->assertEquals('error description', $content['data']['errorDescription']);
        $this->assertEquals(200, $response->getStatusCode());

        $response = Response::serror('error description')->setStatusCode(405);
        $this->assertEquals(405, $response->getStatusCode());
    }
}


src/ResponseFormatter.php
<?php

namespace AdrianoRosa\HttpResponse;

class ResponseFormatter
{

    /**
     * @type array
     */
    protected $response = [
        'code' => 404,
        'status' => 'Not Found',
        'data' => [],
    ];

    /**
     * @param array $data
     * @param int $code
     * @param string $status
     */
    public function __construct(array $data = [], $code = 200, $status = 'success')
    {
        $this->setFormatResponse($data, $code, $status);
    }

    /**
     * @param        $data
     * @param int    $code
     * @param string $status
     *
     * @return $this
     */
    public function setFormatResponse(array $data = [], $code = 200, $status = 'success')
    {
        $this->response['code']   = $code;
        $this->response['status'] = $status;
        $this->response['data']   = $data;

        return $this;
    }

    /**
     * @param mixed $key
     * @param mixed $value
     *
     * @return $this
     */
    public function add($key, $value = null)
    {
        if (is_array($key)) {
            foreach ($key as $k => $v) {
                $this->add($k, $v);
            }
            return $this;
        }

        $this->response[$key] = $value;

        return $this;
    }

    /**
     * Get the response data formatted
     *
     * @return array
     */
    public function getResponse()
    {
        return $this->response;
    }
}


src/Response.php
<?php

namespace AdrianoRosa\HttpResponse;

use \Illuminate\Http\Response as BaseResponse;

/**
 * Class Response
 *
 * @author  Adriano Rosa <http://adrianorosa.com>
 * @date    02/04/16 14:51
 *
 * @package Http
 *
 * @method static|Response json($data = [], $code = null, $headers = [])
 * @method static|Response sjson($data = [], $code = null, $headers = [])
 * @method static|Response error($errorDescription = '', $errorCode = null, $headers = [])
 * @method static|Response serror($errorDescription = '', $errorCode = null, $headers = [])
 */
class Response extends BaseResponse
{
    /**
     * @var array
     */
    protected $append = [];

    /**
     * Factory method for chainability.
     *
     * Example:
     *
     *     return Response::create($body, 200)
     *         ->setSharedMaxAge(300);
     *
     * @param mixed $content The response content, see setContent()
     * @param int   $status  The response status code
     * @param array $headers An array of response headers
     *
     * @return Response
     */
    public static function create($content = '', $status = 200, $headers = array())
    {
        return new static($content, $status, $headers);
    }

    /**
     * Appends data to the response formatter.
     *
     * @param string $key
     * @param mixed $value
     *
     * @return $this
     */
    public function with($key, $value)
    {
        $this->append[$key] = $value;

        return $this;
    }

    /**
     * Format a safe JSON Response.
     *
     * @param array $data
     * @param int $code
     * @param array $headers
     *
     * @return $this
     */
    public function toJson($data = [], $code = null, $headers = [])
    {
        if ( ! is_null($code) ) {
            $this->setStatusCode($code);
        }

        if ( count($data) === 0 ) {
            $data = $this->content;
        }

        if ( $data instanceof \Illuminate\Support\Collection
            || $data instanceof \Illuminate\Database\Eloquent\Model
        ) {
            $data = $data->toArray();
        }

        if ( is_scalar($data) ) {
            $data = empty($data) ? [] : (array) $data;
        }

        $statusText = ($this->statusCode >= 100 && $this->statusCode <= 308) ? 'success' : $this->statusText;

        $formatter = (new ResponseFormatter($data, $this->statusCode, $statusText));

        // append other data to response formatter
        if ( count($this->append) > 0 ) {
            $formatter->add($this->append);
        }

        $default = [
            'Content-Type' => 'application/json'
        ];

        $content = $this->morphToJson($formatter->getResponse());

        $this->setContent($content)->withHeaders(array_merge($default, $headers));

        return $this;
    }

    /**
     * Format a safe JSON Response.
     *
     * @param array $data
     * @param int $code
     * @param array $headers
     *
     * @return $this
     */
    public function safeJson($data = [], $code = null, $headers = [])
    {
        $this->toJson($data, $code, $headers);

        $safeJsonResponse = ")]}',\n" . $this->content;

        $this->setContent($safeJsonResponse);

        return $this;
    }

    /**
     * Return the json error formatted response.
     *
     * @param string $errorDescription
     * @param int $errorCode
     * @param array $headers
     *
     * @return \AdrianoRosa\HttpResponse\Response
     */
    public function jsonError($errorDescription = '', $errorCode = null, $headers = [])
    {
        $statusCode = null;

        if ( array_key_exists($errorCode, self::$statusTexts) ) {
            $statusCode = $errorCode;
        }

        return $this->toJson([
            'errorCode' => $errorCode,
            'errorDescription' => $errorDescription,
        ], $statusCode, $headers);
    }

    /**
     * Return the safe json error formatted response.
     *
     * @param string $errorDescription
     * @param int $errorCode
     * @param array $headers
     *
     * @return \AdrianoRosa\HttpResponse\Response
     */
    public function safeJsonError($errorDescription = '', $errorCode = null, $headers = [])
    {
        $this->jsonError($errorDescription, $errorCode, $headers);

        $this->setContent(")]}',\n" . $this->content);

        return $this;
    }

    /**
     * Magic method to allow easy chaining.
     *
     * @param $method
     * @param $args
     *
     * @return static[]
     */
    public static function __callStatic($method, $args)
    {
        switch ($method) {
            case 'json':
                $function = 'toJson';
                break;

            case 'sjson':
                $function = 'safeJson';
                break;

            case 'error':
                $function = 'jsonError';
                break;

            case 'serror':
                $function = 'safeJsonError';
                break;

            default:
                break;
        }

        if ( isset($function) ) {
            return call_user_func_array([static::create(), $function], $args);
        }

        throw new \BadMethodCallException('Call to undefined method ' . static::class . '::' . $method . '()');
    }
}
